pkg/ovs/openflow/ofctrl_packetin.go

```diff
@@ -40,15 +51,28 @@ func GetTCPHeaderData(ipPkt util.Message) (tcpSrcPort, tcpDstPort uint16, tcpSeq
 		tcpBytes, err = typedIPPkt.Data.(*util.Buffer).MarshalBinary()
 	}
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
-	tcpIn := new(protocol.TCP)
-	err = tcpIn.UnmarshalBinary(tcpBytes)
+	tcpPkt = new(protocol.TCP)
+	err = tcpPkt.UnmarshalBinary(tcpBytes)
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
 
-	return tcpIn.PortSrc, tcpIn.PortDst, tcpIn.SeqNum, tcpIn.AckNum, tcpIn.Code, nil
+	return tcpPkt, nil
+}
+
+func GetTCPDataWithoutOptions(tcpPkt *protocol.TCP) (data []byte, err error) {
+	// TCP.HdrLen is 4-octet unit indicating the length of TCP header including options.
+	tcpOptionsLen := (tcpPkt.HdrLen - tcpStandardHdrLen) * 4
+	// Trim TCP option end mark which has 1 byte for kind, 1 byte for length and no value.
+	if tcpPkt.Data[tcpOptionsLen] == tcpOptionEndKind {
+		tcpOptionsLen += 2
+	}
```

Comment:
I'm not familiar with this, one question: doesn't it already calculate option's length based on HdrLen? why it still needs to move 2 bytes for some cases? Is this (checking 0) the same way kernel extracts payload data?

Reply:
According to rfc793, the end-of-option might be outside the tcpHdrLen(dataOffset), so I add a check here.
https://www.rfc-editor.org/rfc/rfc793#section-3.1:~:text=Specific%20Option%20Definitions